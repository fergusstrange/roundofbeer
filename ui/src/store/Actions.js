const actions = {
  updateError: error => async state => ({
    ...state,
    error,
  }),
  clearError: () => async state => ({
    ...state,
    error: undefined,
  }),
  updateRoundToken: data => async state => ({
    ...state,
    roundToken: data.token,
    round: undefined,
    participatingRounds: Array.isArray(state.participatingRounds)
      ? [...state.participatingRounds.filter(pr => pr.roundUrl !== data.roundUrl),
        {
          roundUrl: data.roundUrl,
          roundToken: data.token,
          participants: data.participants,
        }]
      : [{
        roundUrl: data.roundUrl,
        roundToken: data.token,
        participants: data.participants,
      }],
  }),
  updateRound: round => async state => ({
    ...state,
    round,
  }),
};

export default actions;
