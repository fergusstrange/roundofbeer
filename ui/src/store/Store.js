import { get, set } from 'local-storage';
import { createDakpan } from 'dakpan';
import ApiClient from '../client/Client';

const localStorageKey = 'roundOfBeerLocalStorage';
const client = new ApiClient();

function localStorageStateOrDefaults() {
  return get(localStorageKey)
    || {
      token: undefined,
      round: undefined,
      participant: '',
      participants: [],
    };
}

function updateLocalStore(state) {
  set(localStorageKey, state);
}

const [ContextProvider, useContext] = createDakpan(localStorageStateOrDefaults())({
  clearError: () => async state => ({
    ...state,
    error: undefined,
  }),
  addParticipant: () => ({ participant, ...state }) => ({
    ...state,
    participants: [
      ...state.participants,
      participant,
    ],
    participant: '',
  }),
  updateParticipant: participant => state => ({
    ...state,
    participant,
  }),
  submitParticipants: () => async (state) => {
    const { error, data } = await client.createRound(state.participants);
    return !error
      ? ({
        ...state,
        round: data.round,
        roundToken: data.token,
        participant: '',
        participants: [],
      })
      : ({
        ...state,
        error,
      });
  },
  fetchRound: () => async (state) => {
    const { error, data } = await client.fetchRound(state.roundToken);
    return error
      ? ({
        ...state,
        error,
      })
      : ({
        ...state,
        round: data.round,
        roundToken: data.token,
      });
  },
});

export { ContextProvider, useContext, updateLocalStore };
