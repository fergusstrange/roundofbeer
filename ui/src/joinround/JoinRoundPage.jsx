import React, { Fragment, useState } from 'react';
import { TextField, Grid, Container } from '@material-ui/core';
import Fab from '@material-ui/core/Fab';
import PropTypes from 'prop-types';
import Client from '../client/Client';
import { roundContext } from '../store/Store';

const client = new Client();

export default function JoinRoundPage({ match, history }) {
  const [, actions] = roundContext();
  const [joinRoundPage] = useState({
    nameRef: React.createRef(),
  });

  function submitJoinRound(event) {
    event.preventDefault();
    if (joinRoundPage.nameRef.current.value) {
      client.joinRound(match.params.roundUrl,
        joinRoundPage.nameRef.current.value)
        .then(({ data }) => actions.updateRoundToken(data)
          .then(() => history.push(`/${data.roundUrl}`)))
        .catch(() => actions.updateError('Unable to join that round'));
    }
  }

  return (
    <Fragment>
      <form onSubmit={submitJoinRound}>
        <Grid container spacing={3} direction="column" alignItems="center">
          <Grid item xs={12}>
            <Container>
              <TextField label="Your name" autoFocus inputRef={joinRoundPage.nameRef} />
            </Container>
          </Grid>
          <Grid item xs={12}>
            <Fab
              type="submit"
              variant="extended"
              color="primary"
              aria-label="Add"
            >
              Join Round
            </Fab>
          </Grid>
        </Grid>
      </form>
    </Fragment>
  );
}

JoinRoundPage.propTypes = {
  history: PropTypes.shape().isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      roundUrl: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};
