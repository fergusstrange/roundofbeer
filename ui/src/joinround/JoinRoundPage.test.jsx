import React from 'react';
import renderer from 'react-test-renderer';
import JoinRoundPage from './JoinRoundPage';
import { ContextProvider } from '../store/Store';

describe('Join Round Page Tests', () => {
  it('Renders existing round join page', () => expect(renderer.create(
    <ContextProvider>
      <JoinRoundPage
        history={{}}
        match={{
          params: {
            roundUrl: 'abcdefg',
          },
        }}
      />
    </ContextProvider>,
  ).toJSON())
    .toMatchSnapshot());
});
