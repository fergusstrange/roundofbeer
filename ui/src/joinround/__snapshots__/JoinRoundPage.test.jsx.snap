// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Join Round Page Tests Renders existing round join page 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
          >
            Your name
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={true}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        aria-label="Add"
        className="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-primary"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiFab-label"
        >
          Join Round
        </span>
      </button>
    </div>
  </div>
</form>
`;
